src/traces/main.cpp

```diff
@@ -0,0 +1,164 @@
+        case OperationData::SEARCH:
+            return "SEARCH";
+        case OperationData::DELETE:
+            return "DELETE";
+    }
+    return "UNKNOWN";
+}
+
+
+void
+run_sequential_perf_test()
+{
+    constexpr size_t num_elem = 1000000;
+    SequentialRobinHoodHashTable a;
+
+    for (uint64_t kv = 0; kv < num_elem; ++kv) {
+        a.insert(kv, kv);
+    }
+
+    for (unsigned i = 0; i < 100; ++i) {
+        for (uint64_t kv = 0; kv < num_elem; ++kv) {
+            volatile std::optional<ValueType> r = a.search(kv);
+        }
+    }
+
+    for (uint64_t kv = 0; kv < num_elem; ++kv) {
+        a.remove(kv);
+    }
+}
+
+void *
+threadFunction(void *arg)
+{
+    ThreadData *data = static_cast<ThreadData *>(arg);
+
+    for (size_t i = 0; i < data->numOperations; ++i) {
+        const auto &op = data->operations[i];
+        switch (op.operation) {
+        case OperationData::INSERT:
+            data->hashTable->insert(op.key, op.value);
+            break;
+        case OperationData::SEARCH: {
+            std::pair<ValueType, bool> res = data->hashTable->find(op.key);
+            break;
+        }
+        case OperationData::DELETE:
+            data->hashTable->remove(op.key, op.value);
+            break;
+        default:
+            std::cerr << "Unknown operation: " << operationToString(op.operation) << std::endl;
+            break;
+        }
+    }
+    pthread_exit(nullptr);
+}
+
+
+void
+run_parallel_perf_test()
+{
+    constexpr size_t num_elem = 1000000;
+    constexpr unsigned num_threads = 6; // assumming 6 threads?
+
+    ParallelRobinHoodHashTable a;
+
+    // Create an array of OperationData for each thread
+    OperationData operations[num_elem * 2000000]; 
+
+    // Populate the operations array with a random sequence of operations
+    for (size_t i = 0; i < num_elem * 2000000; ++i) {
+        operations[i].operation = static_cast<OperationData::OperationType>(i % 3);
+        operations[i].key = i % num_elem;
+        operations[i].value = i % num_elem + 1; //just so different
+    }
+
+    // Create thread data and threads
+    pthread_t threads[num_threads];
+    ThreadData threadData[num_threads];
+
+    size_t operations_per_thread = num_elem * 2000000 / num_threads;
+
+    // Launch threads
+    for (unsigned i = 0; i < num_threads; ++i) {
+        threadData[i].hashTable = &a;
+        threadData[i].operations = operations + i * operations_per_thread;
+        threadData[i].numOperations = operations_per_thread;
+        pthread_create(&threads[i], nullptr, threadFunction, &threadData[i]);
+    }
+
+    // Wait for threads to finish
+    for (unsigned i = 0; i < num_threads; ++i) {
+        pthread_join(threads[i], nullptr);
+    }
+}
+
+// operation
+// key val
+//  this array is the trace, now this call funcation that spawns threads that does on the trace
+
+// 1. read trace
+// 2. spawn threads
+// 3. each thread does the operation
+
+int
+main()
+{
+    clock_t start_seq, end_seq;
```

Comment:
You could just reuse these values? Tbh up to you. Don't copy what I did up top haha. Do your own thing. I was just hacking the stuff together 😛